From 9b6851f7221750f8025106840cc3f13ff365793f Mon Sep 17 00:00:00 2001
From: Mehdi Sabwat <mehdi@videolabs.io>
Date: Tue, 27 Apr 2021 15:33:10 +0200
Subject: [PATCH 15/18] core: add initial support for emscripten

posix/sort.c won't be added because qsort_r is not supported.

- add vlc_getProxyUrl stub for emscripten

- implement vlc_thread_id() for emscripten

- add weak attribute support for wasm

Co-Authored-By: Jean-Baptiste Kempf <jb@videolan.org>
---
 include/vlc_common.h     |  2 +-
 src/Makefile.am          | 17 ++++++++++++++++-
 src/emscripten/netconf.c | 39 +++++++++++++++++++++++++++++++++++++++
 src/emscripten/thread.c  | 30 ++++++++++++++++++++++++++++++
 4 files changed, 86 insertions(+), 2 deletions(-)
 create mode 100644 src/emscripten/netconf.c
 create mode 100644 src/emscripten/thread.c

diff --git a/include/vlc_common.h b/include/vlc_common.h
index f62ff86d04..3f64caa455 100644
--- a/include/vlc_common.h
+++ b/include/vlc_common.h
@@ -185,7 +185,7 @@
 # define VLC_USED
 #endif
 
-#if defined (__ELF__) || defined (__MACH__)
+#if defined (__ELF__) || defined (__MACH__) || defined (__wasm__)
 # define VLC_WEAK __attribute__((weak))
 #else
 /**
diff --git a/src/Makefile.am b/src/Makefile.am
index 88bfad5ebb..780c94446c 100644
--- a/src/Makefile.am
+++ b/src/Makefile.am
@@ -442,6 +442,19 @@ libvlccore_la_SOURCES += \
 	posix/timer.c
 endif
 
+if HAVE_EMSCRIPTEN
+libvlccore_la_SOURCES += \
+	posix/thread.c \
+	posix/getaddrinfo.c \
+	posix/error.c \
+	posix/dirs.c \
+	posix/filesystem.c \
+	posix/specific.c \
+	posix/timer.c \
+	emscripten/netconf.c \
+	emscripten/thread.c
+endif
+
 if HAVE_DARWIN
 libvlccore_la_SOURCES += \
 	darwin/error.c \
@@ -485,8 +498,10 @@ if !HAVE_LINUX
 libvlccore_la_SOURCES += posix/wait.c
 endif
 if !HAVE_ANDROID
+if !HAVE_EMSCRIPTEN
+libvlccore_la_SOURCES += posix/sort.c
+endif
 libvlccore_la_SOURCES += \
-	posix/sort.c \
 	posix/thread.c
 if !HAVE_DARWIN
 libvlccore_la_SOURCES += \
diff --git a/src/emscripten/netconf.c b/src/emscripten/netconf.c
new file mode 100644
index 0000000000..f44a11b77d
--- /dev/null
+++ b/src/emscripten/netconf.c
@@ -0,0 +1,39 @@
+/*****************************************************************************
+ * vlc_getProxyUrl for emscripten
+ *****************************************************************************
+ * Copyright (C) 2021 - VideoLabs, VideoLAN and VLC Authors
+ *
+ * This program is free software; you can redistribute it and/or modify it
+ * under the terms of the GNU Lesser General Public License as published
+ * by the Free Software Foundation; either version 2.1 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+ * GNU Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this program; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin Street, Fifth Floor, Boston MA 02110-1301, USA.
+ *****************************************************************************/
+
+#ifdef HAVE_CONFIG_H
+# include "config.h"
+#endif
+#include <strings.h>
+#include <vlc_common.h>
+
+char *vlc_getProxyUrl(const char *url);
+
+/**
+ * Determines the network proxy server to use (if any).
+ * @param url absolute URL for which to get the proxy server
+ * @return proxy URL, NULL if no proxy or error
+ */
+char *vlc_getProxyUrl(const char *url)
+{
+    VLC_UNUSED(url);
+    // TODO: add proxy url to url  
+    return NULL;
+}
diff --git a/src/emscripten/thread.c b/src/emscripten/thread.c
new file mode 100644
index 0000000000..44f5beddfe
--- /dev/null
+++ b/src/emscripten/thread.c
@@ -0,0 +1,30 @@
+/*****************************************************************************
+ * vlc_thread implementation for emscripten
+ *****************************************************************************
+ * Copyright (C) 2021 - VideoLabs, VideoLAN and VLC Authors
+ *
+ * This program is free software; you can redistribute it and/or modify it
+ * under the terms of the GNU Lesser General Public License as published
+ * by the Free Software Foundation; either version 2.1 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+ * GNU Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this program; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin Street, Fifth Floor, Boston MA 02110-1301, USA.
+ *****************************************************************************/
+
+#ifdef HAVE_CONFIG_H
+# include "config.h"
+#endif
+
+#include <pthread.h>
+
+unsigned long vlc_thread_id(void)
+{
+    return pthread_self();
+}
-- 
2.32.0

